gui/blueprint.lua

```diff
@@ -147,6 +147,89 @@ function NamePanel:get_help_pen()
     return self.has_name_collision and COLOR_RED or COLOR_GREY
 end
 
+CycleHotkeyLabel = defclass(CycleHotkeyLabel, widgets.Label)
+CycleHotkeyLabel.ATTRS{
+    key=DEFAULT_NIL,
+    label=DEFAULT_NIL,
+    label_width=DEFAULT_NIL,
+    options=DEFAULT_NIL,
+    option_idx=1,
+    help=DEFAULT_NIL,
+    on_change=DEFAULT_NIL,
+}
+function CycleHotkeyLabel:init()
+    local contents = {
+        {text=self.label, width=self.label_width, key=self.key, key_sep=': ',
+         on_activate=self:callback('cycle')},
+        '  ',
+        {text=function() return self.options[self.option_idx] end},
+    }
+    if self.help then
+        table.insert(contents, '\n')
+        for _,line in ipairs(self.help) do
+            table.insert(contents, {gap=2, text=line, pen=COLOR_GREY})
+            table.insert(contents, '\n')
+        end
+    end
+    self:setText(contents)
+end
+function CycleHotkeyLabel:cycle()
+    if self.option_idx == #self.options then
+        self.option_idx = 1
+    else
+        self.option_idx = self.option_idx + 1
+    end
+    if self.on_change then self.on_change() end
+end
+
+ToggleHotkeyLabel = defclass(ToggleHotkeyLabel, CycleHotkeyLabel)
```

Comment:
Maybe these would be good to move into the widgets library if they're generic. I had implemented something along these lines in https://github.com/lethosor/dfhack/commit/9390ba4da6a3c967f5f370202fa8d922987b5d70 (unmerged)

Reply:
Yours look better designed, I think. Perhaps we could merge this as-is and we can migrate to your implementation once its merged?

Reply:
Ha, thanks, although it could probably stand to be revisited given that I last touched it in 2017. Agreed that this is good for now!